pkg/network/setup/netpod/discover.go

```diff
@@ -134,6 +142,123 @@ func (n NetPod) storePodInterfaceData(vmiSpecIface v1.Interface, ifaceState nmst
+			return err
+		}
+		dhcpConfig.MAC = mac
+
+		linkRoutes, err := filterIPv4RoutesByInterface(currentStatus, podIfaceName)
+		if err != nil {
+			return err
+		}
+		dhcpConfig.Gateway = net.ParseIP(linkRoutes[0].NextHopAddress)
+
+		otherRoutes, err := filterRoutesByNonLocalDestination(linkRoutes, addr)
+		if err != nil {
+			return err
+		}
+
+		dhcpRoutes, err := filterRoutesToDelegateThroughDHCP(otherRoutes)
+		if err != nil {
+			return err
+		}
+		if len(dhcpRoutes) > 0 {
+			dhcpConfig.Routes = &dhcpRoutes
+		}
+	}
+
+	log.Log.V(4).Infof("The generated dhcpConfig: %s\nRoutes: %+v", dhcpConfig.String(), dhcpConfig.Routes)
+	if err := cache.WriteDHCPInterfaceCache(n.cacheCreator, strconv.Itoa(n.podPID), podIfaceName, &dhcpConfig); err != nil {
+		return fmt.Errorf("failed to save DHCP configuration: %w", err)
+	}
+
+	return nil
+}
+
+func filterRoutesToDelegateThroughDHCP(otherRoutes []nmstate.Route) ([]vishnetlink.Route, error) {
+	var dhcpRoutes []vishnetlink.Route
+	for _, nmstateRoute := range otherRoutes {
+		isDefaultRoute := nmstateRoute.Destination == nmstate.DefaultDestinationRoute(vishnetlink.FAMILY_V4).String()
+		var dstAddr *net.IPNet
+		if !isDefaultRoute {
+			_, ipNet, perr := net.ParseCIDR(nmstateRoute.Destination)
+			if perr != nil {
+				return nil, perr
+			}
+			dstAddr = ipNet
+		}
+		route := vishnetlink.Route{
+			Dst: dstAddr,
+			Gw:  net.ParseIP(nmstateRoute.NextHopAddress),
+		}
+		dhcpRoutes = append(dhcpRoutes, route)
+	}
+	return dhcpRoutes, nil
+}
+
+// filterRoutesByNonLocalDestination filters out local routes (the destination is of the local link).
+// Default routes should not be filter out.
+func filterRoutesByNonLocalDestination(linkRoutes []nmstate.Route, addr *vishnetlink.Addr) ([]nmstate.Route, error) {
+	var otherRoutes []nmstate.Route
+	for _, route := range linkRoutes {
+		_, dstIPNet, perr := net.ParseCIDR(route.Destination)
+		if perr != nil {
+			return nil, perr
+		}
+		isDefaultRoute := route.Destination == nmstate.DefaultDestinationRoute(vishnetlink.FAMILY_V4).String()
+		localDestination := dstIPNet.Contains(addr.IP)
+		if isDefaultRoute || !localDestination {
+			otherRoutes = append(otherRoutes, route)
+		}
+	}
+	return otherRoutes, nil
+}
+
+func filterIPv4RoutesByInterface(currentStatus *nmstate.Status, podIfaceName string) ([]nmstate.Route, error) {
+	var linkRoutes []nmstate.Route
+	for _, route := range currentStatus.Routes.Running {
+		ip, _, err := net.ParseCIDR(route.Destination)
+		if err != nil {
+			return nil, err
+		}
+		if isIPv6Family(ip) || isIPv6Family(net.ParseIP(route.NextHopAddress)) {
+			continue
+		}
+		if route.NextHopInterface == podIfaceName {
+			linkRoutes = append(linkRoutes, route)
+		}
+	}
+	if len(linkRoutes) == 0 {
+		return nil, fmt.Errorf("no gateway address found in routes for %s", podIfaceName)
+	}
+	return linkRoutes, nil
+}
+
+func resolveMacAddress(macAddressFromCurrent string, macAddressFromVMISpec string) (net.HardwareAddr, error) {
```

Comment:
The method is bridge binding specific. It's name is confusing and giving the feeling it is a generic method relevant to any binding,
I would move it and all the other bridge binding specific methods to a separate file.